// s3_utilities/index.js
AWS = require("aws-sdk")
var s3 = new AWS.S3();


function S3_utilities(config){
	this._bucket = config.bucket
	this._prefix = config.prefix
}

S3_utilities.prototype.objects_walk = function(previous_result, process_bulk, callback){
	
	self=this
	console.log(self)
	
	iterate("", previous_result, process_bulk, callback)
    
    function iterate(marker, previous_result, process_bulk, cb)
    {
        var params = {
            Bucket: self._bucket, 
            Prefix: self._prefix,
            Marker: marker,
            MaxKeys: 1000
        };
        s3.listObjects(params, function(err, data) {
            if (err) 
                cb(err);
            else
            {
                if(!data.IsTruncated)
                    process_bulk(previous_result, data, cb)
                else
                {
                    process_bulk(previous_result, data, function(err, d){
                        if(err)
                            cb(err)
                        else
                        {
                            iterate(data.Contents[data.Contents.length-1].Key, d, process_bulk, cb)
                        }
                    })
                }
            }
        })        
    }
}


module.exports = S3_utilities;
